{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c81700fe-db85-49d7-b539-4794117dd68c",
   "metadata": {},
   "source": [
    "#### Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "* Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01aa5e14-de31-4e1b-bff0-338f21023454",
   "metadata": {},
   "source": [
    "* When creating a custom exception in a programming language, such as Java or Python, it is important to use the Exception class as the base class for your custom exception. The Exception class provides a set of predefined behaviors and properties that are essential for handling and managing exceptions.\n",
    "* By using the Exception class as the base class for your custom exception, you inherit all the functionality and behavior of the base class. This includes features such as:\n",
    "\n",
    "1. Error Handling: The Exception class provides a mechanism for handling and propagating errors in a structured way. It allows you to catch and handle exceptions at different levels of your program, enabling you to gracefully recover from errors or terminate the program if necessary.\n",
    "\n",
    "2. Exception Propagation: When an exception is thrown, it can be propagated up the call stack until it is caught and handled by an appropriate catch block. The Exception class defines the necessary methods and mechanisms for propagating exceptions and ensuring they are caught and processed correctly.\n",
    "\n",
    "3. Stack Trace: The Exception class includes information about the location and sequence of method calls leading up to the exception. This information is known as the stack trace and is extremely helpful for debugging purposes, as it provides a clear picture of the execution flow and helps identify the root cause of the exception.\n",
    "\n",
    "4. Standard Error Messages: The Exception class provides standard error messages and methods to retrieve and display them. These error messages help developers understand the cause of the exception and facilitate troubleshooting and debugging.\n",
    "\n",
    "5. Compatibility: By using the Exception class, you ensure that your custom exception is compatible with existing exception handling mechanisms in the programming language. This allows other developers to catch and handle your custom exception in a consistent and familiar way."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c4f4491-8c6f-4ee6-b2d2-f95b352e1199",
   "metadata": {},
   "source": [
    "#### Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0e2934ba-4987-4ae9-ba3a-88efa065cd29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BaseException\n",
      "    Exception\n",
      "        TypeError\n",
      "            FloatOperation\n",
      "            MultipartConversionError\n",
      "        StopAsyncIteration\n",
      "        StopIteration\n",
      "        ImportError\n",
      "            ModuleNotFoundError\n",
      "            ZipImportError\n",
      "        OSError\n",
      "            ConnectionError\n",
      "                BrokenPipeError\n",
      "                ConnectionAbortedError\n",
      "                ConnectionRefusedError\n",
      "                ConnectionResetError\n",
      "                    RemoteDisconnected\n",
      "            BlockingIOError\n",
      "            ChildProcessError\n",
      "            FileExistsError\n",
      "            FileNotFoundError\n",
      "            IsADirectoryError\n",
      "            NotADirectoryError\n",
      "            InterruptedError\n",
      "                InterruptedSystemCall\n",
      "            PermissionError\n",
      "            ProcessLookupError\n",
      "            TimeoutError\n",
      "            UnsupportedOperation\n",
      "            itimer_error\n",
      "            herror\n",
      "            gaierror\n",
      "            SSLError\n",
      "                SSLCertVerificationError\n",
      "                SSLZeroReturnError\n",
      "                SSLWantWriteError\n",
      "                SSLWantReadError\n",
      "                SSLSyscallError\n",
      "                SSLEOFError\n",
      "            Error\n",
      "                SameFileError\n",
      "            SpecialFileError\n",
      "            ExecError\n",
      "            ReadError\n",
      "            URLError\n",
      "                HTTPError\n",
      "                ContentTooShortError\n",
      "            BadGzipFile\n",
      "        EOFError\n",
      "            IncompleteReadError\n",
      "        RuntimeError\n",
      "            RecursionError\n",
      "            NotImplementedError\n",
      "                ZMQVersionError\n",
      "                StdinNotImplementedError\n",
      "            _DeadlockError\n",
      "            BrokenBarrierError\n",
      "            BrokenExecutor\n",
      "                BrokenThreadPool\n",
      "            SendfileNotAvailableError\n",
      "            ExtractionError\n",
      "            VariableError\n",
      "        NameError\n",
      "            UnboundLocalError\n",
      "        AttributeError\n",
      "            FrozenInstanceError\n",
      "        SyntaxError\n",
      "            IndentationError\n",
      "                TabError\n",
      "        LookupError\n",
      "            IndexError\n",
      "            KeyError\n",
      "                NoSuchKernel\n",
      "                UnknownBackend\n",
      "            CodecRegistryError\n",
      "        ValueError\n",
      "            UnicodeError\n",
      "                UnicodeEncodeError\n",
      "                UnicodeDecodeError\n",
      "                UnicodeTranslateError\n",
      "            UnsupportedOperation\n",
      "            JSONDecodeError\n",
      "            SSLCertVerificationError\n",
      "            Error\n",
      "            UnsupportedDigestmodError\n",
      "            IllegalMonthError\n",
      "            IllegalWeekdayError\n",
      "            ParserError\n",
      "            ClassNotFound\n",
      "            ClipboardEmpty\n",
      "            MessageDefect\n",
      "                NoBoundaryInMultipartDefect\n",
      "                StartBoundaryNotFoundDefect\n",
      "                CloseBoundaryNotFoundDefect\n",
      "                FirstHeaderLineIsContinuationDefect\n",
      "                MisplacedEnvelopeHeaderDefect\n",
      "                MissingHeaderBodySeparatorDefect\n",
      "                MultipartInvariantViolationDefect\n",
      "                InvalidMultipartContentTransferEncodingDefect\n",
      "                UndecodableBytesDefect\n",
      "                InvalidBase64PaddingDefect\n",
      "                InvalidBase64CharactersDefect\n",
      "                InvalidBase64LengthDefect\n",
      "                HeaderDefect\n",
      "                    InvalidHeaderDefect\n",
      "                    HeaderMissingRequiredValue\n",
      "                    NonPrintableDefect\n",
      "                    ObsoleteHeaderDefect\n",
      "                    NonASCIILocalPartDefect\n",
      "                    InvalidDateDefect\n",
      "            MacroToEdit\n",
      "            InvalidFileException\n",
      "            UnequalIterablesError\n",
      "            InvalidVersion\n",
      "            _InvalidELFFileHeader\n",
      "            InvalidWheelFilename\n",
      "            InvalidSdistFilename\n",
      "            InvalidSpecifier\n",
      "            InvalidMarker\n",
      "            UndefinedComparison\n",
      "            UndefinedEnvironmentName\n",
      "            InvalidRequirement\n",
      "                RequirementParseError\n",
      "            InvalidVersion\n",
      "        AssertionError\n",
      "        ArithmeticError\n",
      "            FloatingPointError\n",
      "            OverflowError\n",
      "            ZeroDivisionError\n",
      "                DivisionByZero\n",
      "                DivisionUndefined\n",
      "            DecimalException\n",
      "                Clamped\n",
      "                Rounded\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Inexact\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Subnormal\n",
      "                    Underflow\n",
      "                DivisionByZero\n",
      "                FloatOperation\n",
      "                InvalidOperation\n",
      "                    ConversionSyntax\n",
      "                    DivisionImpossible\n",
      "                    DivisionUndefined\n",
      "                    InvalidContext\n",
      "        SystemError\n",
      "            CodecRegistryError\n",
      "        ReferenceError\n",
      "        MemoryError\n",
      "        BufferError\n",
      "        Warning\n",
      "            UserWarning\n",
      "                GetPassWarning\n",
      "                FormatterWarning\n",
      "            EncodingWarning\n",
      "            DeprecationWarning\n",
      "                ProvisionalWarning\n",
      "            PendingDeprecationWarning\n",
      "            SyntaxWarning\n",
      "            RuntimeWarning\n",
      "                ProactorSelectorThreadWarning\n",
      "                UnknownTimezoneWarning\n",
      "                PEP440Warning\n",
      "            FutureWarning\n",
      "                ProvisionalCompleterWarning\n",
      "            ImportWarning\n",
      "            UnicodeWarning\n",
      "            BytesWarning\n",
      "            ResourceWarning\n",
      "            DeprecatedTzFormatWarning\n",
      "            PkgResourcesDeprecationWarning\n",
      "        _OptionError\n",
      "        _Error\n",
      "        error\n",
      "        Verbose\n",
      "        Error\n",
      "        SubprocessError\n",
      "            CalledProcessError\n",
      "            TimeoutExpired\n",
      "        TokenError\n",
      "        StopTokenizing\n",
      "        ClassFoundException\n",
      "        EndOfBlock\n",
      "        TraitError\n",
      "        Error\n",
      "        Error\n",
      "            CancelledError\n",
      "            TimeoutError\n",
      "            InvalidStateError\n",
      "        _GiveupOnSendfile\n",
      "        error\n",
      "        Incomplete\n",
      "        TimeoutError\n",
      "        InvalidStateError\n",
      "        LimitOverrunError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        Empty\n",
      "        Full\n",
      "        ArgumentError\n",
      "        ZMQBaseError\n",
      "            ZMQError\n",
      "                ContextTerminated\n",
      "                Again\n",
      "                InterruptedSystemCall\n",
      "            ZMQBindError\n",
      "            NotDone\n",
      "        PickleError\n",
      "            PicklingError\n",
      "            UnpicklingError\n",
      "        _Stop\n",
      "        ArgumentError\n",
      "        ArgumentTypeError\n",
      "        ConfigError\n",
      "            ConfigLoaderError\n",
      "                ArgumentError\n",
      "            ConfigFileNotFound\n",
      "        ConfigurableError\n",
      "            MultipleInstanceError\n",
      "        ApplicationError\n",
      "        error\n",
      "        TimeoutError\n",
      "        error\n",
      "        ReturnValueIgnoredError\n",
      "        KeyReuseError\n",
      "        UnknownKeyError\n",
      "        LeakedCallbackError\n",
      "        BadYieldError\n",
      "        ReturnValueIgnoredError\n",
      "        Return\n",
      "        InvalidPortNumber\n",
      "        error\n",
      "        LZMAError\n",
      "        RegistryError\n",
      "        _GiveupOnFastCopy\n",
      "        Error\n",
      "            NoSectionError\n",
      "            DuplicateSectionError\n",
      "            DuplicateOptionError\n",
      "            NoOptionError\n",
      "            InterpolationError\n",
      "                InterpolationMissingOptionError\n",
      "                InterpolationSyntaxError\n",
      "                InterpolationDepthError\n",
      "            ParsingError\n",
      "                MissingSectionHeaderError\n",
      "        NoIPAddresses\n",
      "        BadZipFile\n",
      "        LargeZipFile\n",
      "        BadEntryPoint\n",
      "        NoSuchEntryPoint\n",
      "        DuplicateKernelError\n",
      "        ErrorDuringImport\n",
      "        NotOneValueFound\n",
      "        CannotEval\n",
      "        OptionError\n",
      "        BdbQuit\n",
      "        Restart\n",
      "        ExceptionPexpect\n",
      "            EOF\n",
      "            TIMEOUT\n",
      "        PtyProcessError\n",
      "        FindCmdError\n",
      "        HomeDirError\n",
      "        ProfileDirError\n",
      "        IPythonCoreError\n",
      "            TryNext\n",
      "            UsageError\n",
      "            StdinNotImplementedError\n",
      "        InputRejected\n",
      "        GetoptError\n",
      "        ErrorToken\n",
      "        PrefilterError\n",
      "        AliasError\n",
      "            InvalidAliasError\n",
      "        Error\n",
      "            InterfaceError\n",
      "            DatabaseError\n",
      "                InternalError\n",
      "                OperationalError\n",
      "                ProgrammingError\n",
      "                IntegrityError\n",
      "                DataError\n",
      "                NotSupportedError\n",
      "        Warning\n",
      "        SpaceInInput\n",
      "        DOMException\n",
      "            IndexSizeErr\n",
      "            DomstringSizeErr\n",
      "            HierarchyRequestErr\n",
      "            WrongDocumentErr\n",
      "            InvalidCharacterErr\n",
      "            NoDataAllowedErr\n",
      "            NoModificationAllowedErr\n",
      "            NotFoundErr\n",
      "            NotSupportedErr\n",
      "            InuseAttributeErr\n",
      "            InvalidStateErr\n",
      "            SyntaxErr\n",
      "            InvalidModificationErr\n",
      "            NamespaceErr\n",
      "            InvalidAccessErr\n",
      "            ValidationErr\n",
      "        ValidationError\n",
      "        EditReadOnlyBuffer\n",
      "        _Retry\n",
      "        InvalidLayoutError\n",
      "        HeightIsUnknownError\n",
      "        ParserSyntaxError\n",
      "        InternalParseError\n",
      "        _PositionUpdatingFinished\n",
      "        SimpleGetItemNotFound\n",
      "        UncaughtAttributeError\n",
      "        HasNoContext\n",
      "        ParamIssue\n",
      "        _JediError\n",
      "            InternalError\n",
      "            WrongVersion\n",
      "            RefactoringError\n",
      "        OnErrorLeaf\n",
      "        InvalidPythonEnvironment\n",
      "        MessageError\n",
      "            MessageParseError\n",
      "                HeaderParseError\n",
      "                BoundaryError\n",
      "            MultipartConversionError\n",
      "            CharsetError\n",
      "        Error\n",
      "        HTTPException\n",
      "            NotConnected\n",
      "            InvalidURL\n",
      "            UnknownProtocol\n",
      "            UnknownTransferEncoding\n",
      "            UnimplementedFileMode\n",
      "            IncompleteRead\n",
      "            ImproperConnectionState\n",
      "                CannotSendRequest\n",
      "                CannotSendHeader\n",
      "                ResponseNotReady\n",
      "            BadStatusLine\n",
      "                RemoteDisconnected\n",
      "            LineTooLong\n",
      "        InteractivelyDefined\n",
      "        KillEmbedded\n",
      "        Error\n",
      "            NoSuchProcess\n",
      "                ZombieProcess\n",
      "            AccessDenied\n",
      "            TimeoutExpired\n",
      "        _Ipv6UnsupportedError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        DebuggerInitializationError\n",
      "        ExpatError\n",
      "        Error\n",
      "            ProtocolError\n",
      "            ResponseError\n",
      "            Fault\n",
      "        ParseBaseException\n",
      "            ParseException\n",
      "            ParseFatalException\n",
      "                ParseSyntaxException\n",
      "        RecursiveGrammarException\n",
      "        ResolutionError\n",
      "            VersionConflict\n",
      "                ContextualVersionConflict\n",
      "            DistributionNotFound\n",
      "            UnknownExtra\n",
      "        _Error\n",
      "        UnableToResolveVariableException\n",
      "        InvalidTypeInArgsException\n",
      "    GeneratorExit\n",
      "    SystemExit\n",
      "    KeyboardInterrupt\n",
      "    CancelledError\n",
      "    AbortThread\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(\" \" * indent + exception_class.__name__)\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "\n",
    "# Start with the base Exception class\n",
    "print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70a772f0-353f-4f25-a3bd-34c46052a50a",
   "metadata": {},
   "source": [
    "#### Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8189a11c-edb2-4bcd-97e4-8148a1a007b8",
   "metadata": {},
   "source": [
    "* The ArithmeticError class is a base class for arithmetic-related errors in Python. It serves as a parent class for several specific error classes that are related to arithmetic operations. Here are two common errors defined in the ArithmeticError class, along with examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bd133b1-257d-4439-ab5d-ff31e3bbf379",
   "metadata": {},
   "source": [
    "1. ZeroDivisionError: This error occurs when attempting to divide a number by zero. It is a subclass of ArithmeticError. Here's an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6652b950-82fd-48f9-b451-eb8b8dd6526a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Division by zero!\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result = 10 / 0  # Attempting to divide by zero\n",
    "except ZeroDivisionError:\n",
    "    print(\"Error: Division by zero!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f06fe14-1ea1-48e3-9d58-e00c2158f9c4",
   "metadata": {},
   "source": [
    "2. OverflowError: This error occurs when the result of an arithmetic operation exceeds the maximum representable value for a numeric type. It is also a subclass of ArithmeticError. Here's an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "65d63a5d-70f9-46ef-9822-b95e31b81ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    result = 10 ** 1000  # Calculating a large power\n",
    "except OverflowError:\n",
    "    print(\"Error: Overflow!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fee05604-4e01-40ef-990f-864e6c913180",
   "metadata": {},
   "source": [
    "#### Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daa329a9-9bee-412c-bf19-351fd8f6b9ea",
   "metadata": {},
   "source": [
    "* The LookupError class is a base class for exceptions that occur when a lookup or indexing operation fails. It serves as a parent class for specific error classes that are related to lookup or indexing operations. The main purpose of the LookupError class is to provide a common base for handling these types of errors in a consistent manner."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2880aabf-cbe1-47c9-96c4-eb0b9f53cef0",
   "metadata": {},
   "source": [
    "* Here are two common errors that are subclasses of LookupError:\n",
    "\n",
    "1. KeyError: This error occurs when trying to access a dictionary or a dictionary-like object using a key that does not exist. Here's an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "61c87a6a-09dd-46bf-9f2c-eec360b2acb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Key not found!\n"
     ]
    }
   ],
   "source": [
    "my_dict = {\"apple\": 1, \"banana\": 2, \"orange\": 3}\n",
    "\n",
    "try:\n",
    "    value = my_dict[\"grape\"]  # Accessing a non-existent key\n",
    "except KeyError:\n",
    "    print(\"Error: Key not found!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6620564e-4ec9-440b-8f4d-1fc442a907a2",
   "metadata": {},
   "source": [
    "2. IndexError: This error occurs when trying to access a sequence (such as a list or a string) using an invalid index. Here's an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d4d42b0-2839-4460-b045-9e95427117c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Index out of range!\n"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3, 4, 5]\n",
    "\n",
    "try:\n",
    "    value = my_list[10]  # Accessing an out-of-range index\n",
    "except IndexError:\n",
    "    print(\"Error: Index out of range!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6842969-3123-4dd3-b42d-b693f59859e3",
   "metadata": {},
   "source": [
    "#### Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00501b95-3c95-45ae-9772-372c4ee53adb",
   "metadata": {},
   "source": [
    "* ImportError is an exception that occurs when an import statement fails to find and load a module. It is raised when there is an issue with importing a module, such as the module not being installed or not being accessible in the current environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4e888d0-3347-4df5-aca8-1b43a028e13e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Module not found or unable to import!\n"
     ]
    }
   ],
   "source": [
    "## Here's an example of ImportError:\n",
    "try:\n",
    "    import non_existent_module\n",
    "except ImportError:\n",
    "    print(\"Error: Module not found or unable to import!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "869c1a14-52c4-4aa9-a595-56b40352b45a",
   "metadata": {},
   "source": [
    "* ModuleNotFoundError is a subclass of ImportError that specifically indicates that a module could not be found. It was introduced in Python 3.6 as a more specific and informative error for cases where a module is not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "27e1cb4c-fdf5-40b8-8e58-35fcc7764e2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Module not found!\n"
     ]
    }
   ],
   "source": [
    "## Here's an example of ModuleNotFoundError:\n",
    "try:\n",
    "    import non_existent_module\n",
    "except ModuleNotFoundError:\n",
    "    print(\"Error: Module not found!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95aacd79-c892-4632-9821-c0b32d31cbb8",
   "metadata": {},
   "source": [
    "#### Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5546137-5daf-4803-b83d-4842a649b943",
   "metadata": {},
   "source": [
    "* Here are some best practices for exception handling in Python:\n",
    "\n",
    "1. Specific Exception Handling: Catch specific exceptions rather than using a generic except block. This allows you to handle different types of exceptions differently and provides more specific error messages. It also helps prevent catching and handling unexpected exceptions.\n",
    "\n",
    "2. Use Multiple Except Blocks: If you need to handle different exceptions differently, use multiple except blocks to handle each exception separately. This allows you to provide specific handling logic for each type of exception.\n",
    "\n",
    "3. Use Finally Block: Use a finally block to specify cleanup code that should be executed regardless of whether an exception occurs or not. This ensures that critical resources are properly released, such as closing files or releasing locks, even if an exception is raised.\n",
    "\n",
    "4. Avoid Broad Except Clauses: Avoid using broad except clauses that catch all exceptions without specifying the type. This can make it difficult to identify and debug specific issues. Instead, catch specific exceptions or use a hierarchy of exception classes.\n",
    "\n",
    "5. Handle Exceptions at the Right Level: Handle exceptions at the appropriate level in your code. This means catching and handling exceptions where you have the necessary information and context to handle them effectively. This can help localize error handling and improve code readability.\n",
    "\n",
    "6. Logging Exceptions: Consider logging exceptions instead of printing error messages directly to the console. Logging provides a more structured and centralized way to capture and review errors, allowing for easier troubleshooting and monitoring.\n",
    "\n",
    "7. Raise Custom Exceptions: When appropriate, raise custom exceptions to provide more meaningful error messages and to encapsulate specific error conditions relevant to your application. Custom exceptions can also help in distinguishing between different types of errors and can be used for more targeted error handling.\n",
    "\n",
    "8. Be Mindful of Performance: Exception handling comes with a performance cost, so avoid using exceptions for regular program flow control. Exceptions should be used for exceptional circumstances, not as a regular part of the program's logic.\n",
    "\n",
    "9. Document Exception Handling: Document the expected exceptions and their meanings in your code's documentation. This helps other developers understand the potential errors and how to handle them correctly.\n",
    "\n",
    "10. Test Exception Scenarios: Create test cases specifically for exception scenarios to ensure that your code behaves as expected when exceptions are raised. This helps identify and fix any issues in the exception handling logic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14307db2-d5f8-4775-ad2c-97bea3fcc8fa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
